rocAL/rocAL/source/mxnet_recordio_reader.cpp

```diff
@@ -0,0 +1,301 @@
+        for (size_t i = 0; i < _batch_count; i++)
+        {
+            _file_names.push_back(_file_names[i - _batch_count]);
+            auto it = _record_properties.find(_file_names[i - _batch_count]);
+            std::tie(_current_file_size, _seek_pos, _data_size_to_read) = it->second;
+            _record_properties.insert(pair<std::string, std::tuple<unsigned int, int64_t, int64_t>>(_file_names[i - _batch_count], std::make_tuple(_current_file_size, _seek_pos, _data_size_to_read)));
+        }
+    }
+}
+
+Reader::Status MXNetRecordIOReader::MXNet_reader()
+{
+    if (_path.find("train") != std::string::npos)
+    {
+        _rec_file = _path + "/train.rec";
+        _idx_file = _path + "/train.idx";
+    }
+    else if(_path.find("val") != std::string::npos)
+    {
+        _rec_file = _path + "/val.rec";
+        _idx_file = _path + "/val.idx";
+    }
+    else
+    {
+        THROW("\nFolder name should be train/val for the train/val MXNet train/validation RecordIO files");
+    }
+    uint rec_size;
+    _file_contents.open(_rec_file, ios::binary);
+    if (!_file_contents)
+        THROW("ERROR: Failed opening the file " + _rec_file);
+    _file_contents.seekg(0, ifstream::end);
+    rec_size = _file_contents.tellg();
+    _file_contents.seekg(0, ifstream::beg);
+
+    ifstream index_file(_idx_file);
+    if(!index_file)
+        THROW("ERROR: Could not open RecordIO index file. Provided path: " + _idx_file);
+
+    while (index_file >> _index >> _offset)
+        _index_list.push_back(_offset);
+    if(_index_list.empty())
+        THROW("ERROR: RecordIO index file doesn't contain any indices. Provided path: " + _idx_file);
+    _index_list.push_back(rec_size);
+    std::sort(_index_list.begin(), _index_list.end());
+    for (size_t i = 0; i < _index_list.size() - 1; ++i)
+        _indices.emplace_back(_index_list[i], _index_list[i + 1] - _index_list[i]);
+    read_image_names();
+    return Reader::Status::OK;
+}
+
+size_t MXNetRecordIOReader::get_file_shard_id()
+{
+    if (_batch_count == 0 || _shard_count == 0)
+        THROW("Shard (Batch) size cannot be set to 0")
+    //return (_file_id / (_batch_count)) % _shard_count;
+    return _file_id  % _shard_count;
+}
+
+void MXNetRecordIOReader::read_image_names()
+{
+    for(int current_index = 0; current_index < (int)_indices.size(); current_index++ )
+    {
+        std::tie(_seek_pos, _data_size_to_read) = _indices[current_index];
+        _file_contents.seekg(_seek_pos, ifstream::beg);
+        _data = new uint8_t[_data_size_to_read];
+        _data_ptr = _data;
+        _file_contents.read((char *)_data_ptr, _data_size_to_read);
+        memcpy(&_magic, _data_ptr, sizeof(_magic));
+        _data_ptr += sizeof(_magic);
+        if(_magic != _kMagic)
+            THROW("ERROR: Invalid MXNet RecordIO: wrong _magic number");
+        memcpy(&_length_flag, _data_ptr, sizeof(_length_flag));
+        _data_ptr += sizeof(_length_flag);
+        _cflag = DecodeFlag(_length_flag);
```

Comment:
Is this needed here? 
Does _cflag needs to be a private member variable? 
I could see it being used locally only.

Reply:
Removed it from private